event/parsing/bootTimeParser.go

```diff
@@ -0,0 +1,247 @@
+	3) Subtract Online birthdate from steps 2 event Birthdate
+	4) Record Metric
+*/
+func (b BootTimeParser) Parse(msg wrp.Message) error {
+	// if event is not an online event, do not continue with calculations
+	if !destinationRegex.MatchString(msg.Destination) || !onlineRegex.MatchString(msg.Destination) {
+		logging.Debug(b.Logger).Log(logging.MessageKey(), "event is not an online event")
+		return nil
+	}
+
+	// get boot time and device id from message
+	bootTimeInt, deviceID, err := getWRPInfo(destinationRegex, msg)
+	if err != nil {
+		return err
+	}
+
+	latestBootTime := time.Unix(bootTimeInt, 0)
+	previousBootTime := int64(0)
+
+	// get events from codex pertaining to this device id
+	events := getEvents(deviceID, b.Logger, b.Address, b.Auth)
+
+	// find the previous boot-time and make sure that the boot time we have is the latest one
+	for _, event := range events {
+		if previousBootTime, err = checkOnlineEvent(event, msg.TransactionUUID, previousBootTime, bootTimeInt); err != nil {
+			logging.Error(b.Logger).Log(logging.MessageKey(), err)
+			if previousBootTime < 0 {
+				// something is wrong with this event's boot time, we shouldn't continue
+				b.UnparsableEventsCount.With(ParserLabel, bootTimeParserLabel, ReasonLabel, eventBootTimeErr).Add(1.0)
+				return nil
+			}
+		}
+	}
+
+	// look through offline events and find the latest offline event
+	latestOfflineEvent := int64(0)
+	for _, event := range events {
+		if latestOfflineEvent, err = checkOfflineEvent(event, previousBootTime, latestOfflineEvent); err != nil {
+			logging.Error(b.Logger).Log(logging.MessageKey(), err)
+		}
+	}
+
+	// boot time calculation
+	restartTime := math.Abs(time.Unix(0, latestOfflineEvent).Sub(latestBootTime).Seconds())
+
+	// add to metrics or log the error
+	if latestOfflineEvent != 0 && previousBootTime != 0 {
+		b.BootTimeHistogram.With(HardwareLabel, msg.Metadata[hardwareKey], FirmwareLabel, msg.Metadata[firmwareKey]).Observe(restartTime)
+	} else {
+		logging.Error(b.Logger).Log(logging.MessageKey(), "failed to get restart time")
+	}
+
+	return nil
+}
+
+// Event is the struct that codex query results will be unmarshalled into.
+type Event struct {
+	MsgType         int               `json:"msg_type"`
+	Source          string            `json:"source"`
+	Dest            string            `json:"dest,omitempty"`
+	TransactionUUID string            `json:"transaction_uuid,omitempty"`
+	ContentType     string            `json:"content_type,omitempty"`
+	Metadata        map[string]string `json:"metadata"`
+	Payload         string            `json:"payload,omitempty"`
+	BirthDate       int64             `json:"birth_date"`
+	PartnerIDs      []string          `json:"partner_ids,omitempty"`
+}
+
+type ResponseEvent struct {
+	Data []Event
+}
+
+// query codex for events related to a device
+func getEvents(device string, logger log.Logger, codexAddress string, codexAuth acquire.Acquirer) []Event {
```

Comment:
In the future, if this function grows to have more parameters or a different Parser needs to contact Codex, it might be nice to make a separate `CodexClient` struct that implements an interface that this and other Parsers could use...especially since this function is the only reason the parser needs `codexAddress` and `codexAuth`. 🙂 Not something that needs to be changed now, just something to keep in mind.